db/db.go

```diff
@@ -43,6 +48,67 @@ type DB interface {
 	DropTable(*schema.Schema) error
 }
 
+func withinTxImpl(db DB, beginStrategy func (db DB) (transaction.Transaction, error), fn func(transaction.Transaction) error) error {
+	log.Debug("Within: begin transaction")
+	defer log.Debug("Within: transaction closed")
+
+	var retryCounter = 0
+
+	for {
+		tx, err := beginStrategy(db)
+
+		if err != nil {
+			log.Warning("Within: failed to begin transaction")
+			return err
+		}
+
+		err = fn(tx)
+
+		if !tx.Closed() {
+			tx.Close()
+		}
+
+		if err == nil {
+			return nil
+		}
+
+		log.Debug("Within: inner function failed with error: %s", err)
+
+		if strings.Contains(err.Error(), "Error 1213: Deadlock found when trying to get lock; try restarting transaction") {
```

Comment:
works only for MySQL. i.e. SQLite error message for deadlock is "database is locked"